File: lang/java-plugins/src/main/java/org/unlaxer/tramli/plugins/subflow/GuaranteedSubflowValidator.java
package org.unlaxer.tramli.plugins.subflow;

import org.unlaxer.tramli.FlowDefinition;
import org.unlaxer.tramli.FlowState;

import java.util.LinkedHashSet;
import java.util.Set;

public final class GuaranteedSubflowValidator {
    public <S extends Enum<S> & FlowState, T extends Enum<T> & FlowState> void validate(
            FlowDefinition<S> parent,
            S parentState,
            FlowDefinition<T> subflow,
            Set<Class<?>> guaranteedTypes) {

        Set<Class<?>> available = new LinkedHashSet<>(parent.dataFlowGraph().availableAt(parentState));
        available.addAll(guaranteedTypes);
        Set<Class<?>> requiredAtEntry = subflow.dataFlowGraph().availableAt(subflow.initialState());
        if (!available.containsAll(requiredAtEntry)) {
            Set<Class<?>> missing = new LinkedHashSet<>(requiredAtEntry);
            missing.removeAll(available);
            throw new IllegalStateException("Subflow " + subflow.name() + " is missing guaranteed types at entry: " + missing);
        }
    }
}

File: lang/ts-plugins/src/subflow/guaranteed-subflow-validator.ts
import type { FlowDefinition } from '@unlaxer/tramli';

/**
 * Validates that a subflow's entry requirements are satisfied
 * by the parent flow's available data at a given state plus any guaranteed types.
 */
export class GuaranteedSubflowValidator {
  validate<S extends string, T extends string>(
    parent: FlowDefinition<S>,
    parentState: S,
    subflow: FlowDefinition<T>,
    guaranteedTypes: Set<string>,
  ): void {
    if (!parent.dataFlowGraph || !subflow.dataFlowGraph) {
      throw new Error('DataFlowGraph is required for subflow validation');
    }
    const available = new Set<string>(parent.dataFlowGraph.availableAt(parentState));
    for (const t of guaranteedTypes) available.add(t);

    const initialState = subflow.initialState;
    if (initialState == null) {
      throw new Error(`Subflow ${subflow.name} has no initial state`);
    }
    const requiredAtEntry = subflow.dataFlowGraph.availableAt(initialState);
    const missing: string[] = [];
    for (const req of requiredAtEntry) {
      if (!available.has(req)) missing.push(req);
    }
    if (missing.length > 0) {
      throw new Error(
        `Subflow ${subflow.name} is missing guaranteed types at entry: [${missing.join(', ')}]`,
      );
    }
  }
}
